File: packages/python/examples/gepa_local_optimizer.py
import json

from axllm import AxGEPA, OptimizerEvaluator


class LocalEvaluator(OptimizerEvaluator):
    def evaluate(self, candidate_map, options=None):
        rows = []
        examples = ((options or {}).get("dataset") or {}).get("train") or []
        instruction = candidate_map.get("qa::instruction", "")
        for example in examples:
            quality = 0.9 if "concise" in instruction.lower() else 0.65
            brevity = 0.8
            scalar = (quality + brevity) / 2
            rows.append(
                {
                    "input": example,
                    "prediction": {"answer": "Ax composes typed LLM programs."},
                    "scores": {"quality": quality, "brevity": brevity},
                    "scalar": scalar,
                }
            )
        total = sum(row["scalar"] for row in rows)
        return {"rows": rows, "avg": total / len(rows), "sum": total, "count": len(rows)}


request = {
    "programKind": "axgen",
    "components": [
        {
            "id": "qa::instruction",
            "owner": "qa",
            "kind": "instruction",
            "current": "Answer clearly and concisely.",
        }
    ],
    "dataset": {
        "train": [{"question": "What is Ax?"}, {"question": "Why use typed signatures?"}],
        "validation": [{"question": "Summarize Ax."}],
    },
    "options": {"numTrials": 0, "maxMetricCalls": 8, "seed": 7},
}

artifact = AxGEPA(seed=7).optimize(request, LocalEvaluator())
assert "qa::instruction" in artifact["componentMap"], artifact
print(json.dumps({"componentMap": artifact["componentMap"], "metadata": artifact["metadata"]}, indent=2, sort_keys=True))

File: packages/rust/examples/gepa_local_optimizer.rs
use axllm::{AxGEPA, AxResult, OptimizerEngine};
use serde_json::{json, Value};

fn main() -> AxResult<()> {
    let request = json!({
        "candidate": {
            "qa::instruction": "Answer clearly and concisely."
        },
        "dataset": {
            "train": [{"question": "What is Ax?"}, {"question": "Why use typed signatures?"}],
            "validation": [{"question": "Summarize Ax."}]
        },
        "options": {"numTrials": 0, "maxMetricCalls": 8, "seed": 7}
    });

    let mut engine = AxGEPA::new();
    let artifact = engine.optimize(request, &mut |candidate: Value| {
        let instruction = candidate["candidate"]["qa::instruction"]
            .as_str()
            .unwrap_or_default();
        let quality = if instruction.to_lowercase().contains("concise") {
            0.9
        } else {
            0.65
        };
        let brevity = 0.8;
        Ok(json!({
            "rows": [{
                "prediction": {"answer": "Ax composes typed LLM programs."},
                "scores": {"quality": quality, "brevity": brevity},
                "scalar": (quality + brevity) / 2.0
            }],
            "avg": (quality + brevity) / 2.0,
            "count": 1
        }))
    })?;
    assert_eq!(artifact["artifact"]["kind"], "gepa");
    println!("{}", serde_json::to_string_pretty(&artifact)?);
    Ok(())
}
